lib/ae_mdw/sync/server.ex

```diff
@@ -1,181 +1,353 @@
+  @spec start_link(GenStateMachine.options()) :: GenStateMachine.on_start()
+  def start_link(_opts), do:
+    GenStateMachine.start_link(__MODULE__, [], name: __MODULE__)
 
-  @spec start_link(GenServer.options()) :: GenServer.on_start()
-  def start_link(_opts),
-    do: GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  @spec new_height(height()) :: :ok
+  def new_height(height), do: GenStateMachine.cast(__MODULE__, {:new_height, height})
 
-  @spec start_sync() :: :ok
-  def start_sync, do: GenServer.cast(__MODULE__, :start_sync)
+  @spec fork(height()) :: :ok
+  def fork(height), do: GenStateMachine.cast(__MODULE__, {:fork, height})
 
-  @spec gens_per_min() :: number()
-  def gens_per_min, do: GenServer.call(__MODULE__, :gens_per_min)
+  @spec gens_per_min() :: gens_per_min()
+  def gens_per_min, do: GenStateMachine.call(__MODULE__, :gens_per_min)
 
   @spec syncing?() :: boolean()
-  def syncing?, do: GenServer.call(__MODULE__, :syncing?)
-
-  @spec done_db(State.t(), gens_per_min()) :: :ok
-  def done_db(state, gens_per_min),
-    do: GenServer.cast(__MODULE__, {:done_db, state, gens_per_min})
+  def syncing?, do: GenStateMachine.call(__MODULE__, :syncing?)
 
-  @spec done_mem(State.t(), gens_per_min()) :: :ok
-  def done_mem(state, gens_per_min),
-    do: GenServer.cast(__MODULE__, {:done_mem, state, gens_per_min})
+  @spec restart_sync() :: :ok
+  def restart_sync, do: GenStateMachine.cast(__MODULE__, :restart_sync)
 
   @impl true
+  @spec init([]) :: :gen_statem.init_result(state())
   def init([]) do
     db_store = DbStore.new()
     mem_store = MemStore.new(db_store)
 
-    db_state = State.new(db_store)
+    db_state = State.new()
     mem_state = State.new(mem_store)
 
-    event_handler = EventHandler.init(chain_height(), db_state, mem_state, &spawn_with_monitor/1)
+    state_data = %__MODULE__{
+      chain_height: nil,
+      fork_height: nil,
+      db_state: db_state,
+      mem_state: mem_state,
+      restarts: 0,
+      gens_per_min: 0
+    }
 
-    {:ok, %__MODULE__{syncing?: false, event_handler: event_handler, gens_per_min: 0}}
+    {:ok, :initialized, state_data}
   end
 
   @impl true
-  @spec handle_call(:syncing?, GenServer.from(), t()) :: {:reply, boolean(), t()}
-  def handle_call(:syncing?, _from, %__MODULE__{syncing?: syncing?} = state) do
-    {:reply, syncing?, state}
+  @spec handle_event(:cast, cast_event(), state(), state_data()) ::
+          :gen_statem.event_handler_result(state())
+  @spec handle_event(:call, call_event(), state(), state_data()) ::
+          :gen_statem.event_handler_result(state())
+  @spec handle_event(:info, info_event(), state(), state_data()) ::
+          :gen_statem.event_handler_result(state())
+  @spec handle_event(:internal, internal_event(), state(), state_data()) ::
+          :gen_statem.event_handler_result(state())
+  def handle_event(:cast, {:new_height, chain_height}, :initialized, state_data) do
+    actions = [{:next_event, :internal, :check_sync}]
```

Comment:
A constant (module attribute) would be helpful to know what actions are passed in `:next_state`.

Reply:
Not sure what you mean by this, there's a typespec for all possible states and events. Can you elaborate please?

Reply:
it's a small detail and on a personal taste for bigger values that are repeated I usually do something like:
`@check_sync_next_event [{:next_event, :internal, :check_sync}]`